src/dune/lib.ml

```diff
@@ -1490,89 +1523,119 @@ end = struct
     let+ () = Result.List.iter ~f:(visit ~stack:[] None) libraries in
     List.filter_map ~f:library_is_default libraries
 
-  let closure_with_overlap_checks db ts ~stack:orig_stack ~linking ~variants
-      ~forbidden_libraries =
-    let visited = ref Set.empty in
-    let unimplemented = ref Vlib.Unimplemented.empty in
-    let res = ref [] in
-    let rec loop (implements_via, t) ~stack =
-      if Set.mem !visited t then
+  module Closure = struct
+    type nonrec t =
```

Comment:
This type is the main change this refactoring introduces. It represents the state of the closure and the `visit` function is responsible for traversing a library and all of its dependencies.